source/io/module_javaScriptInvoke.js

```diff
@@ -0,0 +1,157 @@
+// Using a modified UMD module format. Specifically a modified returnExports (no dependencies) version
+(function (root, globalName, factory) {
+    'use strict';
+    var buildGlobalNamespace = function () {
+        var buildArgs = Array.prototype.slice.call(arguments);
+        return globalName.split('.').reduce(function (currObj, subNamespace, currentIndex, globalNameParts) {
+            var nextValue = currentIndex === globalNameParts.length - 1 ? factory.apply(undefined, buildArgs) : {};
+            return currObj[subNamespace] === undefined ? (currObj[subNamespace] = nextValue) : currObj[subNamespace];
+        }, root);
+    };
+
+    if (typeof define === 'function' && define.amd) {
+        // AMD. Register as a named module.
+        define(globalName, [], factory);
+    } else if (typeof module === 'object' && module.exports) {
+        // Node. "CommonJS-like" for environments like Node but not strict CommonJS
+        module.exports = factory();
+    } else {
+        // Browser globals (root is window)
+        buildGlobalNamespace();
+    }
+}(this, 'NationalInstruments.Vireo.ModuleBuilders.assignJavaScriptInvoke', function () {
+    'use strict';
+
+    /* global Map */
+
+    // Static Private Variables (all vireo instances)
+    var ERRORS = {
+        // Shared
+        NO_ERROR: {
+            CODE: 0,
+            MESSAGE: ''
+        },
+
+        kNIUnableToInvokeAJavaScriptFunction: {
+            CODE: 363656, // TO_DO (CGA): We need to generate a new error code
+            MESSAGE: 'Unable to make JavaScript call for function'
+        },
+
+        kNIUnsupportedTypeInJavaScriptInvoke: {
+            CODE: 363657, // TO_DO (CGA): We need to generate a new error code
+            MESSAGE: 'Unsupported parameter type in JavaScript invoke.'
+        },
+
+        kNIUnableToFindFunctionForJavaScriptInvoke: {
+            CODE: 363658, // TO_DO (CGA): We need to generate a new error code
+            MESSAGE: 'Unable to find JavaScript function'
+        }
+    };
+
+    // Keep in synchrony with JavaScriptParameterTypeEnum in JavaScriptInvoke.cpp
+    var JAVASCRIPT_PARAMETER_TYPE = Object.freeze({
```

Comment:
We talked about it briefly but I'm interested in trying to leverage Vireo's existing type representation for specifying these types.

Vireo already knows how to turn its types into strings, maybe we make some JS code to interpret Vireo's representation of types instead of creating a new way to represent them?

If we develop that code it would be great to make it so that the NIType library can read and serialize Vireo's type representation: https://github.com/ni-kismet/data-types

Tagging @gleonoliva in-case this topic is interesting.

Reply:
Also @prestwick maybe we need to look at open sourcing the NIType library as well